### ISSUE TITLE
Conditional Parameters in Trace Serialization

### ISSUE THREAD

**Issue by janvanrijn** [openml/openml-python#504]
Given Hierarchical parameters, a random search like procedure would not always want to go through the burden of specifying values for all hyperparameters (in fact, randomsearchcv does, but one of my spinoff classes, called defaultsearch, does not. And I think for a good reason, although we can discuss the validity of that)

However, when running the following piece of code:
```
    def test_run_and_upload_defaultsearch(self):
        # more complicated, as it can have hierarchical parameters
        defaultsearch = DefaultSearchCV(
            RandomForestClassifier(n_estimators=5),
            [{'max_features': 4}, {'min_samples_leaf': 1}],
            cv=StratifiedKFold(n_splits=2, shuffle=True))
        # The random states for the RandomizedSearchCV is set after the
        # random state of the RandomForestClassifier is set, therefore,
        # it has a different value than the other examples before
        self._run_and_upload(defaultsearch, '12172')
```
I get the following error:
```
Error
Traceback (most recent call last):
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/unittest/case.py", line 605, in run
    testMethod()
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 421, in test_run_and_upload_defaultsearch
    self._run_and_upload(defaultsearch, '12172')
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 340, in _run_and_upload
    random_state_value=rsv)
  File "/home/jan/projects/openml-python/tests/test_runs/test_run_functions.py", line 154, in _perform_run
    avoid_duplicate_runs=openml.config.avoid_duplicate_runs)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 120, in run_flow_on_task
    res = _run_task_get_arffcontent(flow.model, task, add_local_measures=add_local_measures)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 454, in _run_task_get_arffcontent
    add_local_measures=add_local_measures)
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 576, in _run_model_on_fold
    arff_tracecontent.extend(_extract_arfftrace(model, rep_no, fold_no))
  File "/home/jan/projects/openml-python/openml/runs/functions.py", line 644, in _extract_arfftrace
    serialized_value = json.dumps(model.cv_results_[key][itt_no])
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/jan/anaconda3/envs/openml-defaults/lib/python3.6/json/encoder.py", line 180, in default
    o.__class__.__name__)
TypeError: Object of type 'MaskedConstant' is not JSON serializable
```
Because the MaskedArray cv_result_ is now a 2x2 array, with the empty indices as MaskedConstant. I think we should add an if else statement checking for this in the trace serialization. WDYT?

**Comment by mfeurer** [comment]
Does this also happen with the regular scikit-learn grid and random search? According to the code they're also using masked arrays.

**Comment by janvanrijn** [comment]
I think so, but I don't know how to trigger an example like that (i can try to look into gridsearch)

**Comment by janvanrijn** [comment]
I just added an additional PR, removing the default search and replading it by GridSearch. For this to fully work, some additional things need to change (i.e., GridSearch only has masked arrays if you provide it a list of grids)

What do you think of the current state?

**Comment by mfeurer** [comment]
Not sure which PR you mean, could you please provide me with a link?

**Comment by janvanrijn** [comment]
Typo, I wanted to say: additional commit

**Comment by mfeurer** [comment]
Fixed in #505.
